internal/handlers/handler_register_duo_device.go

```diff
@@ -0,0 +1,119 @@
+package handlers
+
+import (
+	"fmt"
+	"net/url"
+	"strings"
+
+	"github.com/authelia/authelia/v4/internal/duo"
+	"github.com/authelia/authelia/v4/internal/middlewares"
+	"github.com/authelia/authelia/v4/internal/utils"
+)
+
+// SecondFactorDuoDevicesGet handler for retrieving available devices and capabilities from duo api.
+func SecondFactorDuoDevicesGet(duoAPI duo.API) middlewares.RequestHandler {
+	return func(ctx *middlewares.AutheliaCtx) {
+		userSession := ctx.GetSession()
+		values := url.Values{}
+		values.Set("username", userSession.Username)
+
+		ctx.Logger.Debugf("Starting Duo PreAuth for %s", userSession.Username)
+
+		result, message, devices, enrollURL, err := DuoPreAuth(duoAPI, ctx)
+		if err != nil {
+			ctx.Error(fmt.Errorf("duo PreAuth API errored: %s", err), messageMFAValidationFailed)
+			return
+		}
+
+		if result == auth {
+			if devices == nil {
+				ctx.Logger.Debugf("No applicable device/method available for Duo user %s", userSession.Username)
+
+				if err := ctx.SetJSONBody(DuoDevicesResponse{Result: enroll}); err != nil {
+					ctx.Error(fmt.Errorf("unable to set JSON body in response"), messageMFAValidationFailed)
+				}
+
+				return
+			}
+
+			if err := ctx.SetJSONBody(DuoDevicesResponse{Result: auth, Devices: devices}); err != nil {
+				ctx.Error(fmt.Errorf("unable to set JSON body in response"), messageMFAValidationFailed)
+			}
+
+			return
+		}
+
+		if result == allow {
+			ctx.Logger.Debugf("Device selection not possible for user %s, because Duo authentication was bypassed - Defaults to Auto Push", userSession.Username)
+
+			if err := ctx.SetJSONBody(DuoDevicesResponse{Result: allow}); err != nil {
+				ctx.Error(fmt.Errorf("unable to set JSON body in response"), messageMFAValidationFailed)
+			}
+
+			return
+		}
+
+		if result == enroll {
+			ctx.Logger.Debugf("Duo user: %s not enrolled", userSession.Username)
+
+			if err := ctx.SetJSONBody(DuoDevicesResponse{Result: enroll, EnrollURL: enrollURL}); err != nil {
+				ctx.Error(fmt.Errorf("unable to set JSON body in response"), messageMFAValidationFailed)
+			}
+
+			return
+		}
+
+		if result == deny {
+			ctx.Logger.Debugf("Duo User not allowed to authenticate: %s", userSession.Username)
+
+			if err := ctx.SetJSONBody(DuoDevicesResponse{Result: deny}); err != nil {
+				ctx.Error(fmt.Errorf("unable to set JSON body in response"), messageMFAValidationFailed)
+			}
+
+			return
+		}
+
+		ctx.Error(fmt.Errorf("duo PreAuth API errored for %s: %s - %s", userSession.Username, result, message), messageMFAValidationFailed)
+	}
+}
+
+// SecondFactorDuoDevicePost update the user preferences regarding Duo device and method.
+func SecondFactorDuoDevicePost(ctx *middlewares.AutheliaCtx) {
+	device := DuoDeviceBody{}
+
+	err := ctx.ParseBody(&device)
+	if err != nil {
+		ctx.Error(err, messageMFAValidationFailed)
+		return
+	}
+
+	if !utils.IsStringInSlice(device.Method, duo.PossibleMethods) {
```

Comment:
So this links back to my comment I was unsure about if it'd be answered later. Looks like we're only supporting push? But in later lines it looks like we support all of them.

Reply:
Same as above.